File: Validation/ValueSchema.cs
using Newtonsoft.Json;

namespace WebCommons.Model
{
    /// <summary>
    /// Defines the properties of a value for validation and parsing reasons.
    /// </summary>
    public class ValueSchema
    {
        /// <summary>
        /// Defines the name of the property that holds the value when writing to an object. Useful for collecting data from inputs.
        /// </summary>
        [JsonProperty("name")]
        public string Name { get; set; } = string.Empty;

        /// <summary>
        /// Defines the name of the property that holds the value when reading from a source. Useful for filling inputs.
        /// </summary>
        [JsonProperty("fill", NullValueHandling = NullValueHandling.Ignore)]
        public string? Fill { get; set; }

        /// <summary>
        /// Defines the label of the value's input.
        /// </summary>
        [JsonProperty("label", NullValueHandling = NullValueHandling.Ignore)]
        public string? Label { get; set; }

        [JsonProperty("type")]
        public string Type { get; set; } = "string";

        [JsonProperty("isRequired")]
        public bool IsRequired { get; set; } = false;

        [JsonProperty("isEnumerable")]
        public bool IsEnumerable { get; set; } = false;

        [JsonProperty("isNullable")]
        public bool IsNullable { get; set; } = false;

        [JsonProperty("min", NullValueHandling = NullValueHandling.Ignore)]
        public object? Min { get; set; } = null;

        [JsonProperty("max", NullValueHandling = NullValueHandling.Ignore)]
        public object? Max { get; set; } = null;

        /// <summary>
        /// Defines the regex the value must conform to be valid.
        /// </summary>
        [JsonProperty("regex", NullValueHandling = NullValueHandling.Ignore)]
        public string? Regex { get; set; } = null;

        /// <summary>
        /// Defines the possible values that the value must conform to be be valid.
        /// </summary>
        [JsonProperty("values", NullValueHandling = NullValueHandling.Ignore)]
        public object[]? Values { get; set; } = null;

        [JsonProperty("default", NullValueHandling = NullValueHandling.Ignore)]
        public object? DefaultValue { get; set; } = null;

        [JsonProperty("const", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<string, object>? Constants { get; set; } = null;

        [JsonProperty("i18n", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<object, object>? I18n { get; set; } = null;

        [JsonProperty("colors", NullValueHandling = NullValueHandling.Ignore)]
        public Dictionary<object, object>? Colors { get; set; } = null;

        [JsonExtensionData]
        public Dictionary<object, object> Data { get; set; } = new();
    }
}

File: Validation/ValueSchema.js
class ValueSchema {
    constructor(props, props2) {
        props = props ?? {};
        props2 = props2 ?? {};

        this.name = props.name || props2.name;
        this.fill = props.fill || props2.fill || this.name;
        this.label = props.label || props2.label || null;
        this.type = props.type || props2.type || 'string';
        this.regex = props.regex || props.regex || null;
        this.isRequired = [props.isRequired, props2.isRequired, false].firstNonNull();
        this.isEnumerable = [props.isEnumerable, props2.isEnumerable, false].firstNonNull();
        this.isNullable = [props.isNullable, props2.isNullable, false].firstNonNull();
        this.min = [props.min, props2.min, null].firstNonNull();
        this.max = [props.max, props2.max, null].firstNonNull();
        this.values = props.values || props2.values || null;
        this.default = props.default || props2.default || null;
        this.i18n = props.i18n || props.i18n || null;
    }

    toOptions() {
        if (!this.values) return [];
        else return this.values.map(v => {
            return {
                name: (this.i18n && this.i18n[v]) ? translate(this.i18n[v]) : v,
                value: v
            }
        });
    }
}